File: templates/templates/php/Tests/security-options/AuthAdditionalTest.php
declare(strict_types=1);

namespace Speakeasy\SecurityOptions\Tests;

use Speakeasy\SecurityOptions\Tests\CommonHelpers;
use Speakeasy\SecurityOptions\Models\Shared;
use PHPUnit\Framework\TestCase;

final class AuthAdditionalTest extends TestCase
{
    public function testGlobalSecurityBasicHttpSuccess(): void
    {
        CommonHelpers::recordTest('auth-basic-http-global-option');

        // Expected to succeed since BasicHTTP takes priority over AccessToken in global security definition
        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                basicHttp: new Shared\BasicHttp(
                    username: 'testUser',
                    password: 'testPass',
                ),
                accessToken: new Shared\AccessToken(
                    accessToken: 'Bearer ignored',
                ),
            ))
            ->build();

        $response = $sdk->auth->globalSecurityOptionBasicHttp();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->basicAuth);
        $this->assertTrue($response->basicAuth->authenticated);
        $this->assertEquals('testUser', $response->basicAuth->user);
    }

    public function testGlobalSecurityFieldsOrdering(): void
    {
        CommonHelpers::recordTest('auth-global-security-option-fields-ordering');

        // When `maintainOpenApiOrder: false` is set, security fields are sorted in alphabetical order.
        // Since the first non-nil field is selected ApiKeyAuth takes precedence over BasicHttp,
        // which is not valid authentication for this endpoint.
        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                apiKeyAuth: new Shared\ApiKeyAuth(
                    apiKeyAuth: 'Bearer test_api_key',
                ),
                basicHttp: new Shared\BasicHttp(
                    username: 'testUser',
                    password: 'testPass',
                ),
            ))
            ->build();

        try {
            $sdk->auth->globalSecurityOptionBasicHttp();
            $this->fail('Expected APIException was not thrown');
        } catch (\Speakeasy\SecurityOptions\Models\Errors\APIException $e) {
            $this->assertEquals(401, $e->getCode());
        }
    }

    public function testHoistedSecurityAccessTokenFirst(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-option-access-token-first');

        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                apiKeyAuth: new Shared\ApiKeyAuth(
                    apiKeyAuth: 'testApiKey',
                ),
                basicHttp: new Shared\BasicHttp(
                    username: 'username',
                    password: 'password',
                ),
                accessToken: new Shared\AccessToken(
                    accessToken: 'Bearer ghp_xxxx',
                ),
            ))
            ->build();

        $response = $sdk->auth->hoistedSecurityOptionAccessTokenFirst();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->tokenAuthResponse);
        $this->assertEquals('Bearer ghp_xxxx', $response->tokenAuthResponse->token);
    }

    public function testHoistedSecurityApiKeyFirst(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-option-api-key-first');

        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                apiKeyAuth: new Shared\ApiKeyAuth(
                    apiKeyAuth: 'testApiKey',
                ),
                basicHttp: new Shared\BasicHttp(
                    username: 'username',
                    password: 'password',
                ),
                accessToken: new Shared\AccessToken(
                    accessToken: 'Bearer ghp_xxxx',
                ),
            ))
            ->build();

        $response = $sdk->auth->hoistedSecurityOptionApiKeyFirst();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->tokenAuthResponse);
        $this->assertEquals('testApiKey', $response->tokenAuthResponse->token);
    }

    public function testHoistedSecurityBasicHttpOnly(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-option-basic-http-only');

        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                apiKeyAuth: new Shared\ApiKeyAuth(
                    apiKeyAuth: 'testApiKey',
                ),
                basicHttp: new Shared\BasicHttp(
                    username: 'testUser',
                    password: 'testPass',
                ),
                accessToken: new Shared\AccessToken(
                    accessToken: 'Bearer ghp_xxxx',
                ),
            ))
            ->build();

        $response = $sdk->auth->hoistedSecurityOptionBasicHttpOnly();

        $this->assertNotNull($response);
        $this->assertEquals(200, $response->httpMeta->response->getStatusCode());
        $this->assertNotNull($response->basicAuth);
        $this->assertTrue($response->basicAuth->authenticated);
        $this->assertEquals('testUser', $response->basicAuth->user);
    }

    public function testHoistedSecurityInvalidOption(): void
    {
        CommonHelpers::recordTest('auth-hoisted-security-invalid-option');

        $sdk = \Speakeasy\SecurityOptions\SDK::builder()
            ->setSecurity(new Shared\Security(
                basicHttp: new Shared\BasicHttp(
                    username: 'username',
                    password: 'password',
                ),
            ))
            ->build();

        try {
            $sdk->auth->hoistedSecurityOptionAccessTokenFirst();
            $this->fail('Expected APIException was not thrown');
        } catch (\Speakeasy\SecurityOptions\Models\Errors\APIException $e) {
            $this->assertEquals(401, $e->getCode());
        }
    }
}

File: templates/templates/ruby/tests/security-options/auth_additional_test.rb
# frozen_string_literal: true
# typed: true

require_relative "../lib/openapi"
require_relative "common_helper_test"

require "minitest/autorun"
require "minitest/focus"

class TestGlobalSecurityBasicHttpSuccess < Minitest::Test
  def test_global_security_basic_http_success
    record_test("auth-basic-http-global-option")

    # Expected to succeed since BasicHTTP takes priority over AccessToken in global security definition
    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        basic_http: OpenApiSDK::Models::Shared::BasicHttp.new(
          username: "testUser",
          password: "testPass"
        ),
        access_token: OpenApiSDK::Models::Shared::AccessToken.new(
          access_token: "Bearer ignored"
        )
      )
    )

    res = sdk.auth.global_security_option_basic_http

    refute_nil(res)
    assert_equal(200, res.status_code)
    refute_nil(res.basic_auth)
    assert(res.basic_auth.authenticated)
    assert_equal("testUser", res.basic_auth.user)
  end
end

class TestGlobalSecurityFieldsOrdering < Minitest::Test
  def test_global_security_fields_ordering
    record_test("auth-global-security-option-fields-ordering")

    # Expected to fail since APIKeyAuth takes priority over BasicHTTP in global security definition
    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        api_key_auth: OpenApiSDK::Models::Shared::ApiKeyAuth.new(
          api_key_auth: "Bearer test_api_key"
        ),
        basic_http: OpenApiSDK::Models::Shared::BasicHttp.new(
          username: "testUser",
          password: "testPass"
        )
      )
    )

    err = assert_raises(OpenApiSDK::Models::Errors::APIError) do
      sdk.auth.global_security_option_basic_http
    end
    assert_equal(401, err.status_code)
  end
end

class TestHoistedSecurityAccessTokenFirst < Minitest::Test
  def test_hoisted_security_access_token_first
    record_test("auth-hoisted-security-option-access-token-first")

    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        api_key_auth: OpenApiSDK::Models::Shared::ApiKeyAuth.new(
          api_key_auth: "testApiKey"
        ),
        access_token: OpenApiSDK::Models::Shared::AccessToken.new(
          access_token: "Bearer ghp_xxxx"
        )
      )
    )

    res = sdk.auth.hoisted_security_option_access_token_first

    refute_nil(res)
    assert_equal(200, res.status_code)
    refute_nil(res.token_auth_response)
    assert_equal("Bearer ghp_xxxx", res.token_auth_response.token)
  end
end

class TestHoistedSecurityApiKeyFirst < Minitest::Test
  def test_hoisted_security_api_key_first
    record_test("auth-hoisted-security-option-api-key-first")

    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        api_key_auth: OpenApiSDK::Models::Shared::ApiKeyAuth.new(
          api_key_auth: "testApiKey"
        ),
        access_token: OpenApiSDK::Models::Shared::AccessToken.new(
          access_token: "Bearer ghp_xxxx"
        )
      )
    )

    res = sdk.auth.hoisted_security_option_api_key_first

    refute_nil(res)
    assert_equal(200, res.status_code)
    refute_nil(res.token_auth_response)
    assert_equal("testApiKey", res.token_auth_response.token)
  end
end

class TestHoistedSecurityBasicHttpOnly < Minitest::Test
  def test_hoisted_security_basic_http_only
    record_test("auth-hoisted-security-option-basic-http-only")

    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        api_key_auth: OpenApiSDK::Models::Shared::ApiKeyAuth.new(
          api_key_auth: "testApiKey"
        ),
        basic_http: OpenApiSDK::Models::Shared::BasicHttp.new(
          username: "testUser",
          password: "testPass"
        ),
        access_token: OpenApiSDK::Models::Shared::AccessToken.new(
          access_token: "Bearer ghp_xxxx"
        )
      )
    )

    res = sdk.auth.hoisted_security_option_basic_http_only

    refute_nil(res)
    assert_equal(200, res.status_code)
    refute_nil(res.basic_auth)
    assert(res.basic_auth.authenticated)
    assert_equal("testUser", res.basic_auth.user)
  end
end

class TestHoistedSecurityInvalidOption < Minitest::Test
  def test_hoisted_security_invalid_option
    record_test("auth-hoisted-security-invalid-option")

    sdk = OpenApiSDK::SDK.new(
      security: OpenApiSDK::Models::Shared::Security.new(
        basic_http: OpenApiSDK::Models::Shared::BasicHttp.new(
          username: "username",
          password: "password"
        )
      )
    )

    err = assert_raises(OpenApiSDK::Models::Errors::APIError) do
      sdk.auth.hoisted_security_option_access_token_first
    end
    assert_equal(401, err.status_code)
  end
end
